xla/service/cpu/tests/onednn_convolution_test.cc

```diff
@@ -32,159 +33,221 @@ limitations under the License.
+    atol = rtol = (dtype == F32) ? 1e-4 : 1e-2;
+    // TODO(intel-tf): Set default value of user_scratchpad to true after
+    // enabling feature
+    user_scratchpad = false;
+    weights_prepacked = false;
+    dtypeString = primitive_util::LowercasePrimitiveTypeName(dtype);
+  }
 
-TEST_F(ConvolutionTest, Simple2DTestF32) {
-  const char* convolution_module_str = R"(
-  HloModule convolution.test.f32
-
-  ENTRY convolution.test.f32 {
-    arg.0 = f32[1,22,22,1] parameter(0)
-    reshape.0 = f32[1,22,22,1] reshape(arg.0)
-    arg.1 = f32[8,8,1,1] parameter(1)
-    reshape.1 = f32[8,8,1,1] reshape(arg.1)
-    convolution.0 = f32[1,11,11,1] convolution(reshape.0, reshape.1), window={size=8x8 stride=2x2 pad=3_3x3_3}, dim_labels=b01f_01io->b01f
-    reshape.2 = f32[1,11,11,1] reshape(convolution.0)
-    tuple.0 = (f32[1,11,11,1]) tuple(reshape.2)
-    ROOT get-tuple-element.0 = f32[1,11,11,1] get-tuple-element(tuple.0), index=0
-  })";
+  void SetUp() override {
+    if (!IsSupportedType(dtype)) {
+      GTEST_SKIP() << "CPU does not support " << dtypeString;
+    }
+  }
 
-  EXPECT_TRUE(RunAndCompare(convolution_module_str, ErrorSpec{1e-4, 1e-4}));
-  MatchOptimizedHlo(convolution_module_str, conv_rewrite_str_);
-}
+  void SetWeightsPrepacked(bool value) { weights_prepacked = value; }
+
+  void SetUserScratchpad(bool value) { user_scratchpad = value; }
+
+  std::string GetOptimizationsString() {
+    return (user_scratchpad || weights_prepacked)
+               ? absl::Substitute(conv_rewrite_optimizations_str_,
+                                  weights_prepacked, user_scratchpad)
+               : "";
+  }
 
-TEST_F(ConvolutionTest, Simple3DTestBF16) {
-  if (!IsSupportedType(PrimitiveType::BF16)) {
-    GTEST_SKIP() << "CPU does not support BF16.";
+  std::string ConvStringWithOptimizations(
+      const std::vector<absl::string_view> fused_ops) {
+    std::ostringstream stream;
+    std::for_each(
+        fused_ops.begin(), fused_ops.end(),
+        [&](const absl::string_view& arg) { stream << "\"" << arg << "\","; });
+    std::string fusions = stream.str();
+    if (fused_ops.size() > 0) {
+      fusions.pop_back();
+      return absl::Substitute(
+          conv_rewrite_str_,
+          absl::Substitute(conv_rewrite_fusions_str_, fusions),
+          GetOptimizationsString());
+    }
+    return absl::Substitute(conv_rewrite_str_, "", GetOptimizationsString());
   }
 
-  const char* convolution_module_str = R"(
-  HloModule convolution.test.bf16
+  // TODO(intel-tf): Remove this and simplify patterns when Elemental BF16 is
+  // fully supported.
+  PrimitiveType PromotedDtype() {
+    // BF16 is promoted to F32 because not all HLO Instructions currently
+    // support BF16 computations. Meanwhile, FP32 and FP16 elementwise
+    // instructions are not promoted and remain unchanged.
+    switch (dtype) {
+      case F16:
+        return F16;
+      case F32:
+      case BF16:
+      default:
+        return F32;
+    }
```

Comment:
Let's simplify this to
```suggestion
  return (dtype == BF16)? F32 : dtype;
```
We can come back to `switch` later once we need a more complex logic.